File: Mage/Map/Cache/CacheOverlay.m
//
//  CacheOverlay.m
//  MAGE
//
//  Created by Brian Osborn on 12/17/15.
//  Copyright © 2015 National Geospatial Intelligence Agency. All rights reserved.
//

#import "CacheOverlay.h"

@interface CacheOverlay ()

@property (strong, nonatomic) NSString * name;
@property (strong, nonatomic) NSString * cacheName;
@property (nonatomic) enum CacheOverlayType type;
@property (nonatomic) BOOL supportsChildren;

@end

@implementation CacheOverlay

-(instancetype) initWithName: (NSString *) name andType: (enum CacheOverlayType) type andSupportsChildren: (BOOL) supportsChildrens{
    return [self initWithName:name andCacheName:name andType:type andSupportsChildren:supportsChildrens];
}

-(instancetype) initWithName: (NSString *) name andCacheName: (NSString *) cacheName andType: (enum CacheOverlayType) type andSupportsChildren: (BOOL) supportsChildrens{
    self = [super init];
    if(self != nil){
        self.name = name;
        self.cacheName = cacheName;
        self.type = type;
        self.supportsChildren = supportsChildrens;
        self.enabled = false;
    }
    return self;
}

-(NSString *) getName{
    return self.name;
}

-(NSString *) getCacheName{
    return self.cacheName;
}

-(enum CacheOverlayType) getType{
    return self.type;
}

-(BOOL) getSupportsChildren{
    return self.supportsChildren;
}

-(NSArray<CacheOverlay *> *) getChildren{
    return [[NSArray alloc] init];
}

-(NSString *) getInfo{
    return nil;
}

-(void) removeFromMap: (MKMapView *) mapView{
    
}

-(NSString *) onMapClickWithLocationCoordinate: (CLLocationCoordinate2D) locationCoordinate andMap: (MKMapView *) mapView{
    return nil;
}

+(NSString *) buildChildCacheNameWithName: (NSString *) name andChildName: (NSString *) childName{
    return [NSString stringWithFormat:@"%@-%@", name, childName];
}

@end

File: Mage/Map/Cache/CacheOverlay.swift
//
//  CacheOverlay.m
//  MAGE
//
//  Created by Brian Osborn on 12/17/15.
//  Copyright © 2015 National Geospatial Intelligence Agency. All rights reserved.
//

@objc enum CacheOverlayType: Int {
    case XYZ_DIRECTORY
    case GEOPACKAGE
    case GEOPACKAGE_TILE_TABLE
    case GEOPACKAGE_FEATURE_TABLE
};

@objc class CacheOverlay: NSObject {
    var enabled: Bool = false
    var expanded: Bool = false
    var added: Bool = false
    var replaced: CacheOverlay?
    var name: String
    var cacheName: String
    var type: CacheOverlayType
    var iconImageName: String?
    var supportsChildren: Bool = false
    var isChild: Bool = false
    
    init(name: String, cacheName: String? = nil, type: CacheOverlayType, supportsChildren: Bool) {
        self.name = name
        self.cacheName = cacheName ?? name
        self.type = type
        self.supportsChildren = supportsChildren
        self.enabled = false
        self.expanded = false
    }
    
    func getChildren() -> [CacheOverlay] {
        []
    }
    
    func getParent() -> CacheOverlay? {
        nil
    }
    
    func getInfo() -> String? {
        nil
    }
    
    func removeFromMap(mapView: MKMapView) {
        
    }
    
    func onMapClick(locationCoordinates: CLLocationCoordinate2D, mapView: MKMapView) -> String? {
        nil
    }
    
    static func buildChildCacheName(name: String, childName: String) -> String {
        "\(name)-\(childName)"
    }
}
